debian/wrmsr.1
.\"
.\" wrmsr(1)
.\"
.\" Copyright (C) 2008 Andres Salomon
.TH wrmsr 1 "Mar 2008"
.SH NAME
wrmsr \- tool for writing CPU machine specific registers (MSR)
.SH SYNOPSIS
.B "wrmsr [options] regno value"
.SH DESCRIPTION
.B wrmsr 
is a tool used for writing values to a CPU's machine specific registers (MSR).
.br
Note: if you're running a Debian kernel, be sure that the msr.ko kernel
module is loaded. 'modprobe msr' should do the trick. Otherwise, you'll
get an error about
.B wrmsr
not being able to open files in /dev/cpu/...
.SH OPTIONS
.TP
.B --help,          -h
Print a list of available options
.TP
.B --version,       -V
Print current version
.TP
.B --all,           -a
All processors
.TP
.B --processor <#>, -p
Select processor number (default: 0)
.TP
.BR
.SH SEE ALSO
.BR rdmsr(1)
.SH AUTHOR
.br
.B wrmsr 
was written by Transmeta Corporation

This man page was contributed by Andres Salomon <dilinger@debian.org>
for the Debian GNU/Linux system (but may be used by others).



debian/changelog
msr-tools (1.3-3) unstable; urgency=low

  * Merge fix to check that register is valid (closes: #790075). 
  * Update package to catch up on 5 years of new debian policy,
    standards, debhelper, etc.
  * Add a watch file.

 -- Andres Salomon <dilinger@debian.org>  Sun, 02 Dec 2018 07:44:39 +0000

msr-tools (1.3-2) unstable; urgency=low

  * Remove cpuid program; there's already a separate cpuid package in
    Debian that is more fully-featured.  If there's a need to bring back
    the cpuid binary, let me know (closes: #726057).

 -- Andres Salomon <dilinger@debian.org>  Tue, 22 Oct 2013 08:25:00 +0000

msr-tools (1.3-1) unstable; urgency=low

  * New upstream release (closes: #725930).
  * Update the copyright file.
  * Update manpages for new args and new cpuid program.
  * Update standards-version, and add hardening flags to build.
  * Add a note to manpages about ensuring the correct kernel modules
    are loaded when running commands (closes: #721825).

 -- Andres Salomon <dilinger@debian.org>  Thu, 10 Oct 2013 23:39:23 +0000

msr-tools (1.2-3) unstable; urgency=low

  * Only build for i386 and amd64 (closes: #555542).

 -- Andres Salomon <dilinger@debian.org>  Mon, 23 Nov 2009 05:32:50 +0000

msr-tools (1.2-2) unstable; urgency=low

  * Don't allow the package to install anything into /usr/sbin
    (closes: #552958, #554583).
  * Bump the versioned build-dep on debhelper accordingly, as
    override_dh_auto_install is necessary.

 -- Andres Salomon <dilinger@debian.org>  Thu, 05 Nov 2009 18:11:18 +0000

msr-tools (1.2-1) unstable; urgency=low

  * New upstream release (for some value of "new").
  * Bump standards version.
  * Switch from cdbs to debhelper.  It was being too clever in overriding
    our CFLAGS, which resulted in 64bit stuff breaking (closes: #495445).
  * Update copyright info w/ hpa's mark.
  * Drop 001-pwrite_warnings.patch, as it turns out that it was only
    necessary due to the aforementioned CFLAGS clobberage.

 -- Andres Salomon <dilinger@debian.org>  Sun, 25 Oct 2009 14:51:24 +0000

msr-tools (1.1.2-1) unstable; urgency=low

  * Initial release.
  * Create manpages.

 -- Andres Salomon <dilinger@debian.org>  Fri, 07 Mar 2008 12:27:45 +0000


wrmsr.c
/* ----------------------------------------------------------------------- *
 *
 *   Copyright 2000 Transmeta Corporation - All Rights Reserved
 *   Copyright 2004-2008 H. Peter Anvin - All Rights Reserved
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
 *   Boston MA 02110-1301, USA; either version 2 of the License, or
 *   (at your option) any later version; incorporated herein by reference.
 *
 * ----------------------------------------------------------------------- */

/*
 * wrmsr.c
 *
 * Utility to write to an MSR.
 */

#include <errno.h>
#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>
#include <getopt.h>
#include <inttypes.h>
#include <sys/types.h>
#include <dirent.h>
#include <ctype.h>

#include "version.h"

static const struct option long_options[] = {
	{"help",	0, 0, 'h'},
	{"version",	0, 0, 'V'},
	{"all",		0, 0, 'a'},
	{"processor",	1, 0, 'p'},
	{"cpu",		1, 0, 'p'},
	{0, 0, 0, 0}
};
static const char short_options[] = "hVap:";

const char *program;

void usage(void)
{
	fprintf(stderr, "Usage: %s [options] regno value...\n"
		"  --help         -h  Print this help\n"
		"  --version      -V  Print current version\n"
		"  --all          -a  all processors\n"
		"  --processor #  -p  Select processor number (default 0)\n",
		program);
}

void wrmsr_on_cpu(uint32_t reg, int cpu, int valcnt, char *regvals[]);

/* filter out ".", "..", "microcode" in /dev/cpu */
int dir_filter(const struct dirent *dirp)
{
	if (isdigit(dirp->d_name[0]))
		return 1;
	else
		return 0;
}

void wrmsr_on_all_cpus(uint32_t reg, int valcnt, char *regvals[])
{
	struct dirent **namelist;
	int dir_entries;

	dir_entries = scandir("/dev/cpu", &namelist, dir_filter, 0);
	while (dir_entries--) {
		wrmsr_on_cpu(reg, atoi(namelist[dir_entries]->d_name),
				valcnt, regvals);
		free(namelist[dir_entries]);
	}
	free(namelist);
}

int main(int argc, char *argv[])
{
	uint32_t reg;
	int c;
	int cpu = 0;
	unsigned long arg;
	char *endarg;

	program = argv[0];

	while ((c = getopt_long(argc, argv, short_options,
				long_options, NULL)) != -1) {
		switch (c) {
		case 'h':
			usage();
			exit(0);
		case 'V':
			fprintf(stderr, "%s: version %s\n", program,
				VERSION_STRING);
			exit(0);
		case 'a':
			cpu = -1;
			break;
		case 'p':
			arg = strtoul(optarg, &endarg, 0);
			if (*endarg || arg > 255) {
				usage();
				exit(127);
			}
			cpu = (int)arg;
			break;
		default:
			usage();
			exit(127);
		}
	}

	if (optind > argc - 2) {
		/* Should have at least two arguments */
		usage();
		exit(127);
	}

	reg = strtoul(argv[optind++], &endarg, 0);
	if (*endarg) {
		printf("Failed to parse register number.\n");
		exit(127);
	}

	if (cpu == -1) {
		wrmsr_on_all_cpus(reg, argc - optind, &argv[optind]);
	} else {
		wrmsr_on_cpu(reg, cpu, argc - optind, &argv[optind]);
	}

	exit(0);
}

void wrmsr_on_cpu(uint32_t reg, int cpu, int valcnt, char *regvals[])
{
	uint64_t data;
	int fd;
	char msr_file_name[64];

	sprintf(msr_file_name, "/dev/cpu/%d/msr", cpu);
	fd = open(msr_file_name, O_WRONLY);
	if (fd < 0) {
		if (errno == ENXIO) {
			fprintf(stderr, "wrmsr: No CPU %d\n", cpu);
			exit(2);
		} else if (errno == EIO) {
			fprintf(stderr, "wrmsr: CPU %d doesn't support MSRs\n",
				cpu);
			exit(3);
		} else {
			perror("wrmsr: open");
			exit(127);
		}
	}

	while (valcnt--) {
		data = strtoull(*regvals++, NULL, 0);
		if (pwrite(fd, &data, sizeof data, reg) != sizeof data) {
			if (errno == EIO) {
				fprintf(stderr,
					"wrmsr: CPU %d cannot set MSR "
					"0x%08"PRIx32" to 0x%016"PRIx64"\n",
					cpu, reg, data);
				exit(4);
			} else {
				perror("wrmsr: pwrite");
				exit(127);
			}
		}
	}

	close(fd);

	return;
}
